File: searchlib/src/vespa/searchlib/expression/or_predicate_node.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "or_predicate_node.h"
#include "resultnode.h"

#include <algorithm>

namespace search::expression {

using vespalib::Deserializer;
using vespalib::Serializer;

IMPLEMENT_IDENTIFIABLE_NS2(search, expression, OrPredicateNode, MultiArgPredicateNode);

OrPredicateNode::OrPredicateNode() noexcept = default;

OrPredicateNode::~OrPredicateNode() = default;

OrPredicateNode::OrPredicateNode(const OrPredicateNode&) = default;

OrPredicateNode& OrPredicateNode::operator=(const OrPredicateNode&) = default;

bool OrPredicateNode::allow(const DocId docId, const HitRank rank) {
    return std::ranges::any_of(args(), [docId, rank](auto arg){ return arg->allow(docId, rank); });
}

bool OrPredicateNode::allow(const document::Document& doc, const HitRank rank) {
    return std::ranges::any_of(args(), [&doc, rank](auto arg){ return arg->allow(doc, rank); });
}

} // namespace search::expression

File: searchlib/src/main/java/com/yahoo/searchlib/expression/OrPredicateNode.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.expression;

import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

import java.util.List;

/**
 * Protocol class for {@code OrPredicate}.
 *
 * @author johsol
 */
public class OrPredicateNode extends MultiArgPredicateNode {

    public static final int classId = registerClass(0x4000 + 176, OrPredicateNode.class, OrPredicateNode::new);

    public OrPredicateNode() {}

    public OrPredicateNode(List<FilterExpressionNode> args) {
        super(args);
    }

    @Override
    public FilterExpressionNode clone() {
        return new OrPredicateNode(getArgs().map(list -> list.stream()
                .map(FilterExpressionNode::clone).toList()).orElse(null));
    }
    
    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
    }

    @Override
    public boolean equals(Object o) {
        return super.equals(o);
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }
}
